Add generic n-ary product shrinking

sevanspowell opened the pull request:
- Translate the round-robin shrinking algorithm to work on n-ary products (types
  with one constructor [product types], with n constructor arguments), in the
  Generics.SOP universe.
- Write a property to assert that the old selectionParams shrinker and the new
  one produce the same result.
- Replace all instances of liftShrinkN with a generic alternative.

See the following resource for more info on generics-sop: https://raw.githubusercontent.com/kosmikus/cufp-tutorial-2016/master/LectureNotes.pdf

sevanspowell committed:
Write generic round robin shrinker & use it to shrinkSelectionParams

- Translate the round-robin shrinking algorithm to work on n-ary products (types
  with one constructor [product types], with n constructor arguments), in the
  Generics.SOP universe.
- Derive an instance of Generics.SOP.Generic for SelectionParams.
- Replace shrinkSelectionParams with new version that uses generic shrinker
  function ("groundRobinShrink").
- Write a property to assert that the old selectionParams shrinker and the new
  one produce the same result.

sevanspowell committed:
Replace liftShrinkN functions with generic alternative

Commit:
Regenerate nix

sevanspowell committed:
Delete unused code

jonathanknowles committed:
Fix indentation to match project coding style.

Use  4n    space indents for expressions.
Use (4n+2) space indents for `where` clauses.

jonathanknowles committed:
Add generic shrinking operators.

These operators allow us to define shrinkers with a syntax that is very
similar to the applicative syntax for defining generators.

For example:

```haskell

genMyRecord :: Gen MyRecord
genMyRecord = MyRecord
    <$> genFoo
    <*> genBar
    <*> genBaz

shrinkMyRecord :: MyRecord -> [MyRecord]
shrinkMyRecord = genericRoundRobinShrink
    <@> shrinkFoo
    <:> shrinkBar
    <:> shrinkBaz
    <:> Nil

```

jonathanknowles committed:
Add property and unit tests for `interleaveRoundRobin`.

jonathanknowles committed:
Add property tests for `genericRoundRobinShrink`.

Commit:
Regenerate nix.

jonathanknowles committed:
Make `genInteger` a little more interesting.

We give it an increased possibility of generating zero values.

This increases the probability that we'll generate a minimal test record.

jonathanknowles reviewed (approved):
Thanks for making this PR @sevanspowell!

LGTM.

jonathanknowles committed:
Add `libSodiumPostInstall` to `components.tests.unit.postInstall`.

This commit is intended to fix the following Windows build error:

https://hydra.iohk.io/build/8178016/nixlog/4/tail
